File: src/core/runner.ts
import { generateText, LanguageModel } from "ai";
import chalk from "chalk";
import { jsonFormatForLog, logger } from "./logger";
import { getWrappedTools } from "./tool";
import { Agent } from "./types";

const shouldStop = (response: any) => {
  return (
    response?.toolResults?.length === 1 &&
    response?.toolResults[0]?.toolName === "stop"
  );
};

export const run = async (
  agent: Agent,
  model: LanguageModel,
  prompt: string
) => {
  let iteration = 0;
  const { systemPrompt, tools } = agent;
  let shouldStopIteration = false;
  let currentPrompt = `System prompt: ${systemPrompt}\nUser prompt: ${prompt}`;

  logger.info(chalk.bold.cyan("🤖 Starting Dev Agent"));
  logger.info(chalk.bold.blue("📝 System prompt:"), systemPrompt);
  logger.info(chalk.bold.magenta("❓ User prompt:"), prompt);

  while (!shouldStopIteration) {
    iteration++;
    logger.info(
      chalk.yellow.bold(
        `\n🔄 --- Iteration #${iteration} start at ${new Date().toLocaleTimeString()} ---`
      )
    );

    const response = await generateText({
      model,
      prompt: currentPrompt,
      tools: getWrappedTools(tools),
    });

    const shouldStopIteration = shouldStop(response);

    if (shouldStopIteration) {
      const stopResult = response.toolResults[0].result as {
        success: boolean;
        result: string;
      };
      logger.success(chalk.bold.green("✅ Task Complete"));
      logger.info(
        chalk.green(
          `🎯 Success: ${
            stopResult.success
              ? chalk.bold.green("TRUE")
              : chalk.bold.red("FALSE")
          }`
        )
      );
      logger.info(chalk.green("📊 Result:"), stopResult.result);
      break;
    }

    let additionalPrompt = "";
    if (response.text) {
      additionalPrompt = `Agent: ${response.text}`;
      logger.info(chalk.blue.bold("🤖 Agent response:"), response.text.trim());
    }
    if (response.toolResults.length > 0) {
      additionalPrompt += `Tool results: ${JSON.stringify(
        response.toolResults.map((r) => ({
          toolName: r.toolName,
          args: r.args,
          result: r.result,
        }))
      )}`;
      response.toolResults.forEach((r) => {
        logger.info(
          chalk.cyan.bold(`🛠️ Tool ${chalk.italic(r.toolName)}`),
          chalk.gray("called with args:"),
          jsonFormatForLog(r.args),
          chalk.gray("and returned result:"),
          jsonFormatForLog(r.result)
        );
      });
    }

    currentPrompt = `${currentPrompt}\n---\n${additionalPrompt}`;
  }

  logger.close();
};

File: core/runner.py
import asyncio
from typing import Any
from datetime import datetime, timezone
from pydantic_ai import Agent
from pydantic_ai.models.gemini import GeminiModel
from core.agent_config import AgentConfig
from core.logger import logger

class Runner:
    def __init__(self, agent_config: AgentConfig, model: GeminiModel):
        self.agent_config = agent_config
        self.agent = Agent(
            model=model,
            system_prompt=agent_config.system_prompt,
            tools=agent_config.tools,
        )

    async def run(self, prompt: str) -> None:
        logger.info("🤖 Starting Dev Agent")
        logger.info("📝 System prompt:", self.agent_config.system_prompt)
        logger.info("❓ User prompt:", prompt)

        iteration = 0
        async with self.agent.iter(prompt) as agent_run:
            async for node in agent_run:
                # Only increment iteration when we get a model response (LLM call)
                if hasattr(node, 'model_response'):
                    iteration += 1
                    utc_time = datetime.now(timezone.utc).strftime('%H:%M:%S')
                    logger.warning(f"🔄 --- Iteration #{iteration} start at {utc_time} UTC ---")
                    
                    # Handle model responses (tool calls and text)
                    for part in node.model_response.parts:
                        if part.part_kind == 'tool-call':
                            logger.tool_called(part.tool_name, part.args)
                        elif part.part_kind == 'text':
                            logger.info("🤖 Agent response:", part.content.strip())
                
                # Handle tool returns
                elif hasattr(node, 'request'):
                    for part in node.request.parts:
                        if part.part_kind == 'tool-return':
                            logger.tool_returned(part.tool_name, part.content)
                
                # Handle completion
                elif hasattr(node, 'data'):
                    logger.success("✅ Task Complete")
                    logger.info("📊 Final Result:", node.data)

        logger.close()

async def run(agent_config: AgentConfig, model: GeminiModel, prompt: str) -> None:
    """Run the agent with the given configuration and prompt"""
    runner = Runner(agent_config, model)
    await runner.run(prompt)
